skiko/src/iosMain/kotlin/org/jetbrains/skiko/redrawer/MetalRedrawer.ios.kt

```diff
@@ -116,86 +87,64 @@ internal class MetalRedrawer(
-        throw UnsupportedOperationException("This should never be called on iOS")
-    }
-
-    private fun drawIfLayerIsShowing() {
-        if (layer.isShowing()) {
-            draw()
-        }
-    }
-
     private fun draw() {
-        // TODO: maybe make flush async as in JVM version.
-        autoreleasepool { //todo measure performance without autoreleasepool
-            if (!isDisposed) {
-                contextHandler.draw()
-            }
+        if (isDisposed) {
+            return
         }
-    }
 
-    fun finishFrame() {
         autoreleasepool {
-            currentDrawable?.let {
-                val commandBuffer = queue.commandBuffer()!!
-                commandBuffer.label = "Present"
-                commandBuffer.presentDrawable(it)
-                commandBuffer.addCompletedHandler {
-                    // Signal work finish, allow a new command buffer to be scheduled
-                    dispatch_semaphore_signal(inflightSemaphore)
-                }
-                commandBuffer.commit()
-                currentDrawable = null
+            val (width, height) = metalLayer.drawableSize.useContents {
+                width.roundToInt() to height.roundToInt()
             }
-        }
-    }
-}
 
-internal class MetalLayer : CAMetalLayer {
-    private lateinit var skiaLayer: SkiaLayer
-    private lateinit var contextHandler: ContextHandler
-
-    @OverrideInit
-    constructor() : super()
-
-    @OverrideInit
-    constructor(layer: Any) : super(layer)
-
-    fun init(
-        skiaLayer: SkiaLayer,
-        contextHandler: ContextHandler,
-        theDevice: MTLDeviceProtocol
-    ) {
-        this.skiaLayer = skiaLayer
-        this.contextHandler = contextHandler
-        this.setNeedsDisplayOnBoundsChange(true)
-        this.removeAllAnimations()
-        // TODO: looks like a bug in K/N interop.
-        this.device = theDevice as objcnames.protocols.MTLDeviceProtocol?
-        this.pixelFormat = MTLPixelFormatBGRA8Unorm
-        this.contentsGravity = kCAGravityTopLeft
-        doubleArrayOf(0.0, 0.0, 0.0, 0.0).usePinned {
-            this.backgroundColor =
-                CGColorCreate(CGColorSpaceCreateDeviceRGB(), it.addressOf(0))
-        }
-        this.framebufferOnly = false
-        this.opaque = false // For UIKit interop through a "Hole"
-        skiaLayer.view?.let {
-            this.frame = it.frame
-            it.layer.addSublayer(this)
-        }
-    }
+            if (width <= 0 || height <= 0) {
+                return@autoreleasepool
+            }
 
-    fun dispose() {
-        this.removeFromSuperlayer()
-        // TODO: anything else to dispose the layer?
-    }
+            dispatch_semaphore_wait(inflightSemaphore, DISPATCH_TIME_FOREVER)
+
+            val metalDrawable = metalLayer.nextDrawable()
+
+            if (metalDrawable == null) {
+                dispatch_semaphore_signal(inflightSemaphore)
```

Comment:
Do we need to log some warning/error here?

Reply:
Logged warning